File: app/src/main/java/com/example/testproject/util/FileUtil.java
package com.example.testproject.util;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.support.v4.content.FileProvider;
import android.util.Log;

import java.io.*;

/**
 * Created by mike.li.
 */

public class FileUtil {

    public static int dp2px(Context context, float dpValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    public static int px2dp(Context context, float pxValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (pxValue / scale + 0.5f);
    }
    //屏幕的高
    public static int getDisplayHeight(Context context) {
        return Math.max(context.getResources().getDisplayMetrics().widthPixels,
                context.getResources().getDisplayMetrics().heightPixels);
    }
    //屏幕的宽
    public static int getDisplayWidth(Context context) {
        return Math.min(context.getResources().getDisplayMetrics().widthPixels, context
                .getResources().getDisplayMetrics().heightPixels);
    }

    public static float getDisplayDensity(Context context){
        return context.getResources().getDisplayMetrics().scaledDensity;
    }

    public static String getSavePicturePath(){
        return FileUtil.createPath(Environment.getExternalStorageDirectory().getAbsolutePath()+
                File.separator+"irenshi"+ File.separator+"picture"+ File.separator);
    }
    public static String getSaveFilePath(){
        return FileUtil.createPath(Environment.getExternalStorageDirectory().getAbsolutePath()+
                File.separator+"irenshi"+ File.separator+"file"+ File.separator);
    }

    public static String getSaveDBPath(){
        return FileUtil.createPath(Environment.getExternalStorageDirectory().getAbsolutePath()+
                File.separator+"irenshi"+ File.separator+"db"+ File.separator);
    }

    public static boolean checkSdCard() {
        return Environment.getExternalStorageState().equals(
                Environment.MEDIA_MOUNTED);
    }

    public static boolean isExistFile(File file) {
        return file!=null && file.exists();
    }

    public static boolean isExistFile(String file) {
        return CheckUtils.isNotEmpty(file) && new File(file).exists();
    }


    public static void compressImage(File file, File compressFile){
        if(file.length()<1024*1024){
            FileUtil.copy(file.getPath(), compressFile.getPath());
            return;
        }
        Bitmap bitmap = BitmapFactory.decodeFile(file.getPath());
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        // 质量压缩方法，这里100表示不压缩，把压缩后的数据存放到baos中,这里从90开始
        int quality = 90;
        bitmap.compress(Bitmap.CompressFormat.JPEG, quality, baos);
        Log.d("ttt","before=="+baos.toByteArray().length);
        //  循环判断如果压缩后图片是否大于maxSize(例如:100kb),大于继续压缩
        while (baos.toByteArray().length > 1024*1024) {
            baos.reset(); // 重置baos即清空baos
            quality -= 10;// 每次都减少10
            // 这里压缩options，把压缩后的数据存放到baos中
            bitmap.compress(Bitmap.CompressFormat.JPEG, quality, baos);
        }
        Log.d("ttt","after=="+baos.toByteArray().length);
        try{
            FileOutputStream fos = new FileOutputStream(compressFile.getPath());
            fos.write(baos.toByteArray());
            fos.flush();
            fos.close();
        }catch (Exception e){
            e.printStackTrace();
        }
    }

    public static void saveBitmap(File file, Bitmap bitmap) {
        file.deleteOnExit();
        FileOutputStream fos = null;
        try {
            fos = new FileOutputStream(file);
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            bitmap.compress(Bitmap.CompressFormat.WEBP, 100, outputStream);
            fos.write(outputStream.toByteArray());
            fos.flush();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (fos != null) {
                try {
                    fos.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static String createPath(String path) {
        File file = new File(path);
        if (file.exists()) return path;
        boolean mkdirs = file.mkdirs();
        if (mkdirs) {
            return path;
        } else {
            return null;
        }
    }

    public static void copy(String oldPath, String newPath) {
        try {
            FileInputStream fileInputStream = new FileInputStream(oldPath);    //读入原文件
            FileOutputStream fileOutputStream = new FileOutputStream(newPath);
            byte[] buffer = new byte[1024];
            int byteRead;
            while ((byteRead = fileInputStream.read(buffer)) != -1) {
                fileOutputStream.write(buffer, 0, byteRead);
            }
            fileInputStream.close();
            fileOutputStream.flush();
            fileOutputStream.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static Uri getFileUri(Context context, File file){
        Uri fileUri;
        if (Build.VERSION.SDK_INT >= 24) {
            fileUri = FileProvider.getUriForFile(context, "com.ihr.PROVIDER", file);
        } else {
            fileUri = Uri.fromFile(file);
        }
        return fileUri;
    }
}

File: app/src/main/java/com/example/testproject/util/FileUtil.kt
package com.example.testproject.util

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.support.v4.content.FileProvider
import android.util.Log

import java.io.*

/**
 * Created by mike.li.
 */

object FileUtil {

    val savePicturePath: String?
        get() = FileUtil.createPath(
            Environment.getExternalStorageDirectory().absolutePath +
                    File.separator + "irenshi" + File.separator + "picture" + File.separator
        )
    val saveFilePath: String?
        get() = FileUtil.createPath(
            Environment.getExternalStorageDirectory().absolutePath +
                    File.separator + "irenshi" + File.separator + "file" + File.separator
        )

    val saveDBPath: String?
        get() = FileUtil.createPath(
            Environment.getExternalStorageDirectory().absolutePath +
                    File.separator + "irenshi" + File.separator + "db" + File.separator
        )

    fun dp2px(context: Context, dpValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (dpValue * scale + 0.5f).toInt()
    }

    fun px2dp(context: Context, pxValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (pxValue / scale + 0.5f).toInt()
    }

    //屏幕的高
    fun getDisplayHeight(context: Context): Int {
        return Math.max(
            context.resources.displayMetrics.widthPixels,
            context.resources.displayMetrics.heightPixels
        )
    }

    //屏幕的宽
    fun getDisplayWidth(context: Context): Int {
        return Math.min(
            context.resources.displayMetrics.widthPixels, context
                .resources.displayMetrics.heightPixels
        )
    }

    fun getDisplayDensity(context: Context): Float {
        return context.resources.displayMetrics.scaledDensity
    }

    fun checkSdCard(): Boolean {
        return Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED
    }

    fun isExistFile(file: File?): Boolean {
        return file != null && file.exists()
    }

    fun isExistFile(file: String): Boolean {
        return CheckUtils.isNotEmpty<Any>(file) && File(file).exists()
    }


    fun compressImage(file: File, compressFile: File) {
        val bitmap = BitmapFactory.decodeFile(file.path)
        val baos = ByteArrayOutputStream()
        // 质量压缩方法，这里100表示不压缩，把压缩后的数据存放到baos中,这里从90开始
        var quality = 90
        bitmap.compress(Bitmap.CompressFormat.JPEG, quality, baos)
        Log.d("ttt", "before==" + baos.toByteArray().size)
        //  循环判断如果压缩后图片是否大于maxSize(例如:100kb),大于继续压缩
        while (baos.toByteArray().size > 1024 * 1024) {
            baos.reset() // 重置baos即清空baos
            quality -= 10// 每次都减少10
            // 这里压缩options，把压缩后的数据存放到baos中
            bitmap.compress(Bitmap.CompressFormat.JPEG, quality, baos)
        }
        Log.d("ttt", "after==" + baos.toByteArray().size)
        try {
            val fos = FileOutputStream(compressFile.path)
            fos.write(baos.toByteArray())
            fos.flush()
            fos.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun saveBitmap(file: File, bitmap: Bitmap) {
        file.deleteOnExit()
        var fos: FileOutputStream? = null
        try {
            fos = FileOutputStream(file)
            val outputStream = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.WEBP, 100, outputStream)
            fos.write(outputStream.toByteArray())
            fos.flush()
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            if (fos != null) {
                try {
                    fos.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        }
    }

    fun createPath(path: String): String? {
        val file = File(path)
        if (file.exists()) return path
        val mkdirs = file.mkdirs()
        return if (mkdirs) {
            path
        } else {
            null
        }
    }

    fun getFileUri(context: Context, file: File): Uri {
        val fileUri: Uri
        if (Build.VERSION.SDK_INT >= 24) {
            fileUri = FileProvider.getUriForFile(context, "com.ihr.PROVIDER", file)
        } else {
            fileUri = Uri.fromFile(file)
        }
        return fileUri
    }
}